### ISSUE TITLE
Feature: determine if publishing is credential-based or OIDC-based

### ISSUE THREAD

**Issue by di** [ossf/scorecard#2761]
**Is your feature request related to a problem? Please describe.**
GitHub Actions supports OIDC identities for workflow runs: https://docs.github.com/en/actions/deployment/security-hardening-your-deployments/about-security-hardening-with-openid-connect

Some ecosystems now support OIDC-based publishing with these identities, e.g. for PyPI: https://github.com/pypi/warehouse/issues/12465 (currently in closed beta), which can be identified based on the usage of the official action without supplying credentials: https://github.com/pypa/gh-action-pypi-publish#publishing-with-openid-connect

This should be considered more secure than using long-lived credentials like API tokens or username/passwords for publishing. It also has the added benefit of linking the source repository to the published artifact in a verifiable way.

**Describe the solution you'd like**
For ecosystems that support OIDC-based publishing, Scorecard should penalize repositories that use long-lived credentials for publishing, and reward projects that adopt OIDC-based publishing, either via a new check or an existing check.

**Comment by di** [comment]
As an example, here's an instance of the change that would be necessary to move from less-secure long-lived credentials to more secure OIDC-based publishing: https://github.com/sigstore/sigstore-python/pull/566/

**Comment by pnacht** [comment]
Would this be a new metric for the Packaging check or a new check in itself? It seems a good fit for Packaging, but then that raises the issue of how to score projects in different ecosystems (those that support OIDC and those that don't).

As a separate check this becomes easy to handle: if a project belongs to an OIDC-friendly ecosystem, it gets scored. Otherwise, the check is ignored.

The same thing could be done within Packaging, but it might seem odd to the maintainer that a Python project gets an 8/10 for publishing (without OIDC) while an R project gets a 10/10 for the exact same thing.

**Comment by laurentsimon** [comment]
+1 from me. Great to encourage users to use OIDC instead of long-lived tokens.

**Comment by varunsh-coder** [comment]
JFYI - I have created an [issue](https://github.com/step-security/secure-repo/issues/2032) in https://github.com/step-security/secure-repo to automate transformation of GitHub Actions workflows to use OIDC.

**Comment by di** [comment]
More examples of this already available other ecosystems: https://dart.dev/tools/pub/automated-publishing

**Comment by laurentsimon** [comment]
I'm +1 to have this under Packaging check. Folks who have a 10 today will drop to a lower score (say, 8) and it will encourage them to update.

**Comment by woodruffw** [comment]
Not stale 🙂

**Comment by woodruffw** [comment]
Not stale. Is there a way to disable the stale check on this issue? I believe it can't really go "stale" per se, since it's purely a feature request.

**Comment by pnacht** [comment]
> Not stale. Is there a way to disable the stale check on this issue? I believe it can't really go "stale" per se, since it's purely a feature request.

Yes, the Action has "exempt labels" for issues and PRs. However, Scorecard currently only has that for "bug", "good first issue" and something else. Maybe should add a "never-stale" label.

**Comment by woodruffw** [comment]
Makes sense. I think "good first issue" would also make sense in this specific case :slightly_smiling_face:

**Comment by laurentsimon** [comment]
Here's another one we should detect https://blog.rubygems.org/2023/12/14/trusted-publishing.html
